ambient.go

```diff
@@ -0,0 +1,156 @@
+// without placing constrainsts on programs under test.
+type ambientPlugin struct {
+	// Short name of the provider, such as "eks" or "aws".
+	Provider string
+
+	// Desired version, such as "5.42.0". Only one of Version, LocalPath should be set.
+	Version string
+
+	// Local path to a plugin binary, such as "../bin/pulumi-resource-eks".
+	LocalPath string
+}
+
+// Builds a PATH environment variable value suitable for setting up the environment with the desired
+// ambient plugins. Auto-installed via Pulumi CLI.
+func pathWithAmbientPlugins(t *testing.T, originalPATH string, plugins ...ambientPlugin) (string, error) {
+	_, err := ensurePluginsInstalled(plugins)
+	if err != nil {
+		return "", err
+	}
+	newPaths := []string{}
+	for _, p := range plugins {
+		var newPath string
+		if p.LocalPath != "" {
+			newPath = filepath.Dir(p.LocalPath)
+		} else if p.Version != "" {
+			pi := pluginInfo{
+				Name:    p.Provider,
+				Kind:    "resource",
+				Version: p.Version,
+			}
+			var err error
+			newPath, err = pluginPath(pi)
+			if err != nil {
+				return "", err
+			}
+		}
+		newPaths = append(newPaths, newPath)
+	}
+	oldPaths := strings.Split(originalPATH, string(os.PathListSeparator))
+	allPaths := append(newPaths, oldPaths...)
+
+	for _, np := range newPaths {
+		t.Logf("Adding to PATH: %s\n", np)
+	}
+
+	return strings.Join(allPaths, string(os.PathListSeparator)), nil
+}
+
+type pluginInfo struct {
+	Name    string `json:"name"`
+	Kind    string `json:"kind"`
+	Version string `json:"version"`
+}
+
+// Find the locally installed Plugin path.
+func pluginPath(info pluginInfo) (string, error) {
+	plugins, err := workspace.GetPluginDir()
+	if err != nil {
+		return "", err
+	}
+	return filepath.Join(plugins, fmt.Sprintf("%s-%s-v%s", info.Kind, info.Name, info.Version)), nil
+}
+
+// Use pulumi plugin install to ensure all requested plugins are installed, if not already.
+func ensurePluginsInstalled(plugins []ambientPlugin) ([]pluginInfo, error) {
+	matched := []pluginInfo{}
+	installed, err := findInstalledPlugins()
+	if err != nil {
+		return nil, err
+	}
+	for _, p := range plugins {
+		if p.Version != "" {
+			isInstalledAlready := false
+			for _, i := range installed {
+				if i.Kind == "resource" && i.Name == p.Provider && i.Version == p.Version {
+					isInstalledAlready = true
+					matched = append(matched, i)
+					break
+				}
+			}
+			if !isInstalledAlready {
+				info := pluginInfo{Kind: "resource", Name: p.Provider, Version: p.Version}
+				if err := installPlugin(info); err != nil {
+					return nil, err
+				}
+				installed = append(installed, info)
+				matched = append(matched, info)
+			}
+		}
+	}
+	return matched, nil
+}
+
+// Call pulumi plugin install.
+func installPlugin(p pluginInfo) error {
+	pulumi, err := exec.LookPath("pulumi")
+	if err != nil {
+		return fmt.Errorf("cannot find pulumi CLI in PATH: %w", err)
+	}
+	cmd := exec.Command(pulumi, "plugin", "install", p.Kind, p.Name, p.Version)
```

Comment:
I would refactor this out into its own `pulumiCmd(args ...string) *exec.Command` function.